File: apps/mobile/targets/watch/Theme.swift
import SwiftUI
import WatchKit

/// DVNT brand on watchOS. True-black canvas (OLED power + contrast) and the
/// token table from docs/dvnt-design-system.md §1 — cyan / violet / magenta,
/// signal, gold. These are the phone's literal hexes, not a watch-local remix:
/// a member glancing from wrist to phone must not see two brands.
enum DVNT {
    static let canvas = Color.black

    // MARK: Brand stops (design-system §1)
    static let cyan = Color(hex: 0x3FDCFF)
    static let violet = Color(hex: 0x8A40CF)
    static let magenta = Color(hex: 0xFF5BFC)
    static let signal = Color(hex: 0xFC253A)
    static let gold = Color(hex: 0xF5C518)

    static let hairline = Color.white.opacity(0.10)
    static let surface = Color.white.opacity(0.04)

    /// Text sitting on *artwork* rather than on the canvas. A flyer is an
    /// unknown, often mid-tone image, so the canvas ramp (`textDim` 0.60,
    /// `textFaint` 0.40) goes muddy on top of it even through a scrim. These
    /// run brighter on purpose — same hierarchy, more headroom.
    enum OnArt {
        static let primary = Color.white
        static let secondary = Color.white.opacity(0.82)
        static let tertiary = Color.white.opacity(0.66)
    }
    /// Between `textDim` and full white. Four call sites had invented 0.70 and
    /// 0.75 for the same job; both round here, and a 0.02–0.03 alpha shift on
    /// white is below the perceptual floor of an OLED watch panel.
    static let textBright = Color.white.opacity(0.72)
    static let textDim = Color.white.opacity(0.60)
    /// The missing rung. Six call sites wanted "dimmer than `textDim`, brighter
    /// than `textFaint`" and each invented its own 0.5–0.55, which is precisely
    /// how a palette drifts.
    static let textMuted = Color.white.opacity(0.55)
    static let textFaint = Color.white.opacity(0.40)

    /// Primary accent. Cyan is stop 1 and the design system's "primary accent".
    static let accent = cyan

    /// **The Deviant Gradient** — `linear-gradient(100deg, #3FDCFF 0%, #8A40CF 52%,
    /// #FF5BFC 100%)`. One brand stroke, and the design system spends it on a short
    /// list only (CTA, price chip, the "going" ring, one hairline). On the wrist
    /// that list is exactly one thing: `AccessRing`. Everything else is flat
    /// `surface` + `hairline` — see the "Cut the glow" note in the design system.
    ///
    /// 100° in CSS is measured clockwise from north, so the gradient line runs
    /// left→right with a slight upward tilt: (sin100°, −cos100°).
    static let brandGradient = LinearGradient(
        stops: [
            .init(color: cyan, location: 0.0),
            .init(color: violet, location: 0.52),
            .init(color: magenta, location: 1.0),
        ],
        startPoint: UnitPoint(x: 0.008, y: 0.413),
        endPoint: UnitPoint(x: 0.992, y: 0.587)
    )

    /// The same three stops swept around a circle, for `AccessRing`. Wrapped
    /// back to cyan at the seam so a rotating sweep has no visible hard edge.
    static let brandSweep = AngularGradient(
        stops: [
            .init(color: cyan, location: 0.00),
            .init(color: violet, location: 0.36),
            .init(color: magenta, location: 0.70),
            .init(color: cyan, location: 1.00),
        ],
        center: .center
    )

    /// Tier accent matching the phone's TIER_ACCENT map.
    static func tierAccent(_ tier: String?) -> Color {
        switch tier {
        case "free": return cyan
        case "vip": return violet
        case "table": return magenta
        default: return Color(hex: 0x34A2DF) // ga
        }
    }
}

// MARK: - Type

extension DVNT {
    /// The type system. Two registers, deliberately different jobs.
    ///
    /// STRUCTURAL type — tier, status, "present at door", counts — is condensed
    /// and tracked. That is venue vernacular: laminates, door signage, marquee
    /// boards. It also buys real horizontal room on a 40mm screen, so a label
    /// like "GENERAL ADMISSION" sets on one line instead of truncating.
    ///
    /// CONTENT type — event names, host messages, times — stays default width.
    /// Condensing content hurts legibility at arm's length in the dark, which is
    /// the one condition this app is actually used in.
    /// Sizes obey the watchOS glanceability floor (HIG W-GL-03): body is never
    /// below 16pt, titles never below 18pt. The previous screens set 10–15pt
    /// across the board — 32 of 42 explicit sizes were under the floor, which is
    /// why nothing read at arm's length in a dark room.
    enum TypeScale {
        /// Stamped structural label — tier, status, "PRESENT AT DOOR", the
        /// marquee eyebrows. Uppercase at the call site; pair with
        /// `.tracking(DVNT.TypeScale.stampTracking)`.
        ///
        /// This is Republica Minor, and it is the one place the watch gets its
        /// own voice: a bold ITALIC display face reads as venue signage —
        /// laminates, door boards, a marquee — which is exactly the register
        /// these labels occupy. Titles stay Space Grotesk, because an italic
        /// display face is wrong for content you actually read (event names,
        /// host messages) at arm's length in a dark room.
        ///
        /// NAME WARNING: the PostScript name is `RepublicaMinor-BoldItalic`,
        /// NOT the `Republica-Minor` filename. Font.custom takes the PostScript
        /// name; UIAppFonts takes the filename. Swapping them fails silently to
        /// the system face — verify with the name table, never by eye.
        static func stamp(_ size: CGFloat = 13) -> Font {
            .custom("RepublicaMinor-BoldItalic", size: size)
        }

        /// Screen and row titles. Floor is 18pt.
        static func title(_ size: CGFloat = 18) -> Font {
            .custom("SpaceGrotesk-Bold", size: max(size, 18))
        }

        /// Body copy: host messages, event names on the pass. Floor is 16pt.
        static func body(_ size: CGFloat = 16) -> Font {
            .custom("Inter-Regular", size: max(size, 16))
        }

        /// Secondary metadata — dates, venue, staleness. Deliberately the one
        /// register allowed under the body floor, so use it only for text the
        /// user never has to read to act.
        static func caption(_ size: CGFloat = 14) -> Font {
            .custom("Inter-Regular", size: size)
        }

        /// Tracking for stamped labels. Applied via `.tracking(DVNT.TypeScale.stampTracking)`.
        ///
        /// Raised from 1.4 when `stamp` became Republica Minor. That face is
        /// bold AND italic with tight native sidebearings, so set solid it
        /// closes up into a block at 13pt on a 40mm panel — the slanted stems
        /// of adjacent caps nearly touch. 2.2pt at 13pt is ~0.17em, which is
        /// heavy for a text face but correct for tracked-out signage caps and
        /// is what gives the marquee its air.
        ///
        /// If a label ever needs a much larger size, prefer size * 0.17 over
        /// this constant — tracking should scale with the type, and every
        /// current call site sits in the 11-14pt band this is tuned for.
        static let stampTracking: CGFloat = 2.2

        /// Countdowns and counts. Monospaced digits so a ticking value does not
        /// reflow the row it sits in — pair with `.contentTransition(.numericText())`.
        static func numeral(_ size: CGFloat = 28) -> Font {
            .custom("SpaceMono-Regular", size: size).monospacedDigit()
        }

        /// A numeral that is *signage* rather than content — the door count, a
        /// capacity board. Condensed and heavy for the same reason `stamp` is
        /// (venue vernacular: laminates, door boards), but monospaced so a live
        /// count does not reflow its row as it ticks.
        ///
        /// Split from `numeral` deliberately: a plain `numeral` here would drop
        /// `.width(.condensed)` and quietly restyle the one screen a host stares
        /// at all night.
        static func numeralStamp(_ size: CGFloat = 34) -> Font {
            .custom("SpaceMono-Regular", size: size).monospacedDigit()
        }

        /// SF Symbol sizing. Icons were the one register still set with raw
        /// `.system(size:)` at the call site (13 / 15 / 28 across four files);
        /// these are those three values, named.
        enum Icon {
            static let inline: CGFloat = 13   // trailing metadata glyphs
            static let row: CGFloat = 15      // leading glyph on a list row
            static let control: CGFloat = 20  // glyph inside a tap target
            static let hero: CGFloat = 28     // empty-state / focal
        }
    }

    /// Surface elevations. Three steps and a hairline — the exact opacities
    /// already in use, lifted out of the call sites that repeated them.
    enum Surface {
        static let low = Color.white.opacity(0.04)    // == DVNT.surface
        static let mid = Color.white.opacity(0.06)    // list row fill
        static let high = Color.white.opacity(0.15)   // inactive badge
        static let hairline = Color.white.opacity(0.08)
    }

    /// Corner radii. `14` was hardcoded at five call sites in EventListView alone.
    enum Radius {
        static let card: CGFloat = 14
        static let chip: CGFloat = 10
        static let hero: CGFloat = 20
    }

    /// Motion vocabulary. One set of curves so nothing reads as ad-hoc.
    ///
    /// Nothing here loops. This target's rule (see `AccessRing`) is that a
    /// repeating animation must be driven by a `TimelineView` the view can
    /// pause — `repeatForever` keeps ticking off-screen and in Always-On,
    /// which is a battery bug you cannot see in the simulator.
    enum Motion {
        static let enter = Animation.spring(response: 0.42, dampingFraction: 0.82)
        static let settle = Animation.spring(response: 0.55, dampingFraction: 0.9)
        static let quick = Animation.easeOut(duration: 0.18)

        /// Staggered list entrance, capped so a long list does not crawl in.
        static func stagger(_ i: Int) -> Animation { enter.delay(min(Double(i) * 0.055, 0.4)) }
    }
}

// MARK: - Haptics

extension DVNT {
    /// Haptic vocabulary — the same gesture must always feel the same. Named by
    /// meaning, not by `WKHapticType`, so the mapping is changed in one place.
    enum Haptic {
        /// Crown-driven page change.
        static func page() { WKInterfaceDevice.current().play(.click) }
        /// Entering a destination — the ticket stack, the door.
        static func enter() { WKInterfaceDevice.current().play(.start) }
        /// A ticket just went checked-in during a sync. The physical "you're in".
        static func admit() { WKInterfaceDevice.current().play(.success) }
        /// A pass that cannot be presented was opened.
        static func blocked() { WKInterfaceDevice.current().play(.failure) }
    }

    /// 4-based spacing scale. Every magic number in this target should resolve here.
    enum Space {
        static let hair: CGFloat = 2
        static let tight: CGFloat = 4
        static let snug: CGFloat = 6
        static let base: CGFloat = 8
        static let roomy: CGFloat = 12
        static let loose: CGFloat = 16
    }
}

extension Color {
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xff) / 255,
            green: Double((hex >> 8) & 0xff) / 255,
            blue: Double(hex & 0xff) / 255,
            opacity: 1
        )
    }
}

File: apps/mobile/wear/src/main/java/com/dvnt/app/wear/Theme.kt
package com.dvnt.app.wear

import android.graphics.Typeface
import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.FiniteAnimationSpec
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.wear.compose.material3.ColorScheme
import androidx.wear.compose.material3.MaterialTheme
import androidx.wear.compose.material3.Typography

/**
 * DVNT brand on Wear OS. The Android half of `apps/mobile/targets/watch/Theme.swift`
 * — same hexes, same floors, same names, so a member glancing from an Apple Watch to
 * a Pixel Watch to the phone never sees two brands.
 *
 * House rule: NOTHING in this module writes a raw `dp` or `sp` at a call site. Every
 * spatial or type value resolves to a token below. If a value is missing, add it here.
 *
 * Design references: `wear-os-design-guidelines`
 *   WR-CC-01 true-black canvas, WR-GL-03 type floors, WR-AM-01/03 ambient discipline.
 */
object Dvnt {

    /** WR-CC-01. Black pixels are free on OLED and buy the contrast for nothing. */
    val canvas = Color.Black

    // ---------------------------------------------------------------- Brand stops
    // docs/dvnt-design-system.md §1 — the phone's literal hexes.
    val cyan = Color(0xFF3FDCFF)
    val violet = Color(0xFF8A40CF)
    val magenta = Color(0xFFFF5BFC)
    val signal = Color(0xFFFC253A)
    val gold = Color(0xFFF5C518)

    /** Primary accent — cyan is stop 1 and the design system's "primary accent". */
    val accent = cyan

    val hairline = Color.White.copy(alpha = 0.10f)
    val textDim = Color.White.copy(alpha = 0.60f)
    val textFaint = Color.White.copy(alpha = 0.40f)

    /**
     * **The Deviant Gradient** — `linear-gradient(100deg, #3FDCFF 0%, #8A40CF 52%,
     * #FF5BFC 100%)`.
     *
     * It is spent on ONE focal element per screen and never behind a list or a
     * section. On the wrist that budget is the ring around the QR plate. Everywhere
     * else is flat [canvas] + [Surface.hairline] — a gradient behind a list lights
     * every pixel of an OLED panel for decoration, and this device is on a battery
     * that also has to survive a night out.
     *
     * 100° in CSS is measured clockwise from north, so the line runs left→right
     * with a slight upward tilt: (sin100°, −cos100°).
     */
    fun brandGradient(widthPx: Float, heightPx: Float): Brush = Brush.linearGradient(
        0.00f to cyan,
        0.52f to violet,
        1.00f to magenta,
        start = Offset(0.008f * widthPx, 0.413f * heightPx),
        end = Offset(0.992f * widthPx, 0.587f * heightPx),
    )

    /** The same three stops swept around a circle, wrapped back to cyan at the seam. */
    fun brandSweep(center: Offset): Brush = Brush.sweepGradient(
        0.00f to cyan,
        0.36f to violet,
        0.70f to magenta,
        1.00f to cyan,
        center = center,
    )

    /** Tier accent, matching the phone's TIER_ACCENT map. */
    fun tierAccent(tier: String?): Color = when (tier) {
        "free" -> cyan
        "vip" -> violet
        "table" -> magenta
        else -> Color(0xFF34A2DF) // ga
    }

    /** Status accent. Only `valid` is brand-coloured; everything else is muted. */
    fun statusAccent(status: TicketStatus): Color = when (status) {
        TicketStatus.VALID -> cyan
        TicketStatus.CHECKED_IN -> gold
        TicketStatus.TRANSFER_PENDING -> violet
        TicketStatus.REVOKED, TicketStatus.EXPIRED, TicketStatus.CANCELLED, TicketStatus.UNKNOWN -> signal
    }

    // ------------------------------------------------------------------- Surfaces
    /** Three steps and a hairline. Flat fills only — see [brandGradient]. */
    object Surface {
        val low = Color.White.copy(alpha = 0.04f)
        val mid = Color.White.copy(alpha = 0.06f)   // list row fill
        val high = Color.White.copy(alpha = 0.15f)  // inactive badge
        val hairline = Color.White.copy(alpha = 0.08f)
        /** The QR plate. A scanner needs a light field; this is the only non-black. */
        val code = Color(0xFFFFFFFF)
    }

    // --------------------------------------------------------------------- Radius
    object Radius {
        val chip: Dp = 10.dp
        val card: Dp = 14.dp
        val hero: Dp = 20.dp
    }

    // ---------------------------------------------------------------------- Space
    /** 4-based scale. Every magic number in this module resolves here. */
    object Space {
        val hair: Dp = 2.dp
        val tight: Dp = 4.dp
        val snug: Dp = 6.dp
        val base: Dp = 8.dp
        val roomy: Dp = 12.dp
        val loose: Dp = 16.dp
        /** Round-display safe inset for a full-width list (WR-RD-01/02). */
        val arc: Dp = 20.dp
    }

    // ---------------------------------------------------------------------- Sizes
    object Size {
        val minTouch: Dp = 48.dp        // WR-AC / Material minimum
        val tierDot: Dp = 10.dp
        val ringStroke: Dp = 3.dp
        val hairlineStroke: Dp = 1.dp
        val qrPlate: Dp = 116.dp        // the focal element on the detail screen
        val qrQuietZone: Dp = 8.dp
        val iconInline: Dp = 13.dp      // trailing metadata glyph
        val iconRow: Dp = 15.dp         // leading glyph on a list row
        val iconHero: Dp = 28.dp        // empty state / focal
    }

    // ----------------------------------------------------------------------- Type
    /**
     * Two registers with different jobs, ported 1:1 from `Theme.swift`.
     *
     * STRUCTURAL type — tier, status, "PRESENT AT DOOR", counts — is condensed and
     * tracked: venue vernacular (laminates, door signage), and it buys real
     * horizontal room on a 41mm round screen where "GENERAL ADMISSION" otherwise
     * truncates.
     *
     * CONTENT type — event names, times, venues — stays default width. Condensing
     * content hurts legibility at arm's length in the dark, which is the only
     * condition this app is ever used in.
     *
     * THE FLOORS ARE HARD (WR-GL-03, and the watchOS target's W-GL-03): body is
     * never below 16sp, titles never below 18sp. `caption` at 14sp is the single
     * register allowed under the body floor and is reserved for text nobody has to
     * read in order to act.
     */
    object Type {
        /** Android's stock condensed face. Present on every Wear OS system image. */
        val condensed: FontFamily = FontFamily(Typeface.create("sans-serif-condensed", Typeface.NORMAL))

        private val BODY_FLOOR = 16.sp
        private val TITLE_FLOOR = 18.sp

        /** Tracking for stamped labels. */
        val stampTracking = 1.4.sp

        /** Stamped structural label — tier, status, "PRESENT AT DOOR". Uppercase at
         *  the call site. 13sp condensed+heavy reads as signage, not body copy. */
        val stamp = TextStyle(
            fontFamily = condensed,
            fontSize = 13.sp,
            lineHeight = 15.sp,
            fontWeight = FontWeight.Black,
            letterSpacing = stampTracking,
        )

        /** Screen and row titles. Floor 18sp. */
        val title = TextStyle(
            fontSize = TITLE_FLOOR,
            lineHeight = 22.sp,
            fontWeight = FontWeight.SemiBold,
        )

        /** Body copy: event names on the pass, host messages. Floor 16sp. */
        val body = TextStyle(
            fontSize = BODY_FLOOR,
            lineHeight = 20.sp,
            fontWeight = FontWeight.Normal,
        )

        /** Secondary metadata — dates, venue, staleness. Never load-bearing. */
        val caption = TextStyle(
            fontSize = 14.sp,
            lineHeight = 17.sp,
            fontWeight = FontWeight.Normal,
        )

        /** Countdowns and counts. Monospaced digits so a changing value never
         *  reflows the row it sits in. */
        val numeral = TextStyle(
            fontFamily = FontFamily.Monospace,
            fontSize = 28.sp,
            lineHeight = 32.sp,
            fontWeight = FontWeight.Bold,
        )
    }

    // --------------------------------------------------------------------- Motion
    /**
     * One vocabulary so nothing reads as ad-hoc.
     *
     * NOTHING HERE REPEATS. A `repeatForever` animation keeps ticking off-screen and
     * in ambient, which is a battery bug you cannot see in an emulator (WR-AM-03).
     * If a looping affordance is ever needed, drive it from a state the screen can
     * stop, and stop it in `onEnterAmbient`.
     */
    object Motion {
        fun <T> enter(): FiniteAnimationSpec<T> =
            spring(dampingRatio = 0.82f, stiffness = Spring.StiffnessMediumLow)

        fun <T> settle(): FiniteAnimationSpec<T> =
            spring(dampingRatio = 0.90f, stiffness = Spring.StiffnessLow)

        fun <T> quick(): FiniteAnimationSpec<T> =
            tween(durationMillis = 180, easing = CubicBezierEasing(0.2f, 0f, 0f, 1f))

        /** Staggered list entrance, capped so a long list does not crawl in. */
        fun staggerMillis(index: Int): Int = minOf(index * 55, 400)
    }
}

private val DvntColors = ColorScheme(
    primary = Dvnt.cyan, primaryDim = Dvnt.cyan,
    secondary = Dvnt.violet, tertiary = Dvnt.magenta,
    background = Dvnt.canvas, surfaceContainer = Dvnt.Surface.mid,
    surfaceContainerLow = Dvnt.Surface.low, surfaceContainerHigh = Dvnt.Surface.high,
    error = Dvnt.signal, onPrimary = Color.Black, onSecondary = Color.White,
    onBackground = Color.White, onSurface = Color.White, onSurfaceVariant = Dvnt.textDim,
)
private val DvntTypography = Typography(
    titleLarge = Dvnt.Type.title, titleMedium = Dvnt.Type.title, titleSmall = Dvnt.Type.title,
    bodyLarge = Dvnt.Type.body, bodyMedium = Dvnt.Type.body, bodySmall = Dvnt.Type.caption,
    labelLarge = Dvnt.Type.body, labelMedium = Dvnt.Type.caption, labelSmall = Dvnt.Type.stamp,
    numeralLarge = Dvnt.Type.numeral,
)
@Composable
fun DvntWearTheme(content: @Composable () -> Unit) {
    MaterialTheme(colorScheme = DvntColors, typography = DvntTypography, content = content)
}
